keps/sig-node/compute-device-assignment.md

```diff
@@ -0,0 +1,169 @@
+
+### Guaranteed Device Scheduling
+
+* As a _Cluster Admin and Application Developer_, I have a workload with strict requirements about the set of devices it is assigned that cannot be expressed with current kubernetes APIs (e.g. numa topology requirements). If my workload is assigned to a node that has the devices I need, but not in the correct configuration, my workloads performance suffers considerably.  Thus, the node's best-effort attempt to provide a "good" set of devices is not enough to meet the needs of my workload.  I can run a custom scheduler to assign devices to containers during scheduling, and can _guarantee_ that the set of devices the workload is assigned meets my requirements.
+
+## Alternatives Considered
+
+* Kubelet `/devices` endpoint: Add a new alpha endpoint to the kubelet's read-only port containing devices and their associated pods and containers.
+
+* Add a field to pod status rather than pod spec.
+
+## Changes
+
+### API Changes
+A new object `ComputeDevice` will be defined and a new variable `ComputeDevices` will be added in `Container` object which will represent a list of `ComputeDevice` objects.
+
+NOTE: `ComputeDevices` will be added as an alpha field and will be behind a feature-gate, `ComputeDevices`  
+
+```golang
+// ComputeDevice describes the devices assigned to this container for a given ResourceName
+type ComputeDevice struct {
+	// DeviceIDs is the list of devices assigned to this container
+	DeviceIDs []string
+	// ResourceName is the name of the compute resource
+	ResourceName string
+}
+
+// Container represents a single container that is expected to be run on the host.
+type Container struct {
+    ...
+	// ComputeDevices contains the devices assigned to this container
+	// This field is alpha-level and is only honored by servers that enable the ComputeDevices feature.
+	// +optional
+	ComputeDevices []ComputeDevice
+	...
+}
+```
+
+To be able to initialize/update `ComputeDevices`, a new verb, `BindResource()`, will be defined for binding `ResourceBinding` objects with a pod object:
+```golang
+// ResourceBinding ties a resource to an object.
+type ResourceBinding struct {
+	metav1.TypeMeta
+	// ObjectMeta describes the object that is being bound.
+	// +optional
+	metav1.ObjectMeta
+
+	// Bindings describes the resource that the object should be bound to
+	// +optional
+	Bindings []ComputeDeviceBinding
+}
+
+type ComputeDeviceBinding struct {
+	// ContainerName is the name of the container being bound
+	// +optional
+	ContainerName string
+
+	// ComputeDevices is the devices being bound to the container
+	// +optional
+	ComputeDevices []ComputeDevice
+}
+```
+`client-go` will be extended to support `BindResource` verb/subresource and also changes will be made in `pkg/registry/core` to handle `POST`ing of `ResourceBinding` object and eventually updating `ComputeDevices` in the pod spec's containers.
```

Comment:
Just curious, what are the use cases for updating after the initial binding? This seems like something that may or may not work for various classes of compute device.

Reply:
Or does this update refer to the step of copying the compute resources from the binding to the container?

Reply:
This update refers to the step of copying the compute resources from the binding to the container.  We should probably validate that devices cannot be removed (and thus cannot be changed to a different device).